shell/common/renderer_context_manager.h

```diff
@@ -0,0 +1,119 @@
+  using SetCurrentContextBlock =
+      std::function<bool(std::shared_ptr<RendererContext>)>;
+  // Should return a `nullptr` if there is no current context.
+  using GetCurrentContextBlock =
+      std::function<std::shared_ptr<RendererContext>()>;
+
+ public:
+  //------------------------------------------------------------------------------
+  /// Switches the renderer context to the flutter's contexts.
+  ///
+  /// In construction, it should set the current context to a flutter's context
+  /// In destruction, it should reset the current context.
+  ///
+  class RendererContextSwitch {
+   public:
+    RendererContextSwitch(RendererContextManager& manager,
+                          std::shared_ptr<RendererContext> context)
+        : manager_(manager) {
+      bool result = manager_.PushContext(context);
+      has_pushed_context_ = true;
+      result_ = result;
+    }
+
+    ~RendererContextSwitch() {
+      if (!has_pushed_context_) {
+        return;
+      }
+      manager_.PopContext();
+    }
+
+    // Returns true if the context switching was successful.
+    bool GetResult() { return result_; }
+
+   private:
+    RendererContextManager& manager_;
+    bool result_;
+    bool has_pushed_context_;
+
+    FML_DISALLOW_COPY_AND_ASSIGN(RendererContextSwitch);
+  };
+
+  RendererContextManager(std::shared_ptr<RendererContext> context,
+                         std::shared_ptr<RendererContext> resource_context,
+                         SetCurrentContextBlock set_current_block,
+                         GetCurrentContextBlock get_current_block)
+      : context_(context),
+        resource_context_(resource_context),
+        set_current_context_block_(set_current_block),
+        get_current_context_block_(get_current_block) {}
+  ~RendererContextManager() = default;
+
+  //----------------------------------------------------------------------------
+  /// @brief      Make the flutter's context as current context.
+  ///
+  /// @return     A `RendererContextSwitch` which ensures the current context is
+  /// the `context` that used in the constructor.
+  std::unique_ptr<RendererContextSwitch> MakeCurrent() {
+    return std::make_unique<RendererContextSwitch>(*this, context_);
+  }
+
+  //----------------------------------------------------------------------------
+  /// @brief      Make the flutter's resources context as current context.
+  ///
+  /// @return     A `RendererContextSwitch` which ensures the current context is
+  /// the `resource_context` that used in the constructor.
+  std::unique_ptr<RendererContextSwitch> ResourceMakeCurrent() {
+    return std::make_unique<RendererContextSwitch>(*this, resource_context_);
+  }
+
+  //----------------------------------------------------------------------------
+  /// @return     The flutter context.
+  std::shared_ptr<RendererContext> GetContext() { return context_; }
+
+ private:
+  std::shared_ptr<RendererContext> context_;
```

Comment:
Sorry. Did not submit this one last comment in the previous review. Using a shared pointer here puts the added burden of making sure the render context can be destroyed on any thread. In reality, render context cannot are tied to specific threads.